orangecontrib/text/widgets/owpreprocess.py

```diff
@@ -1,70 +1,706 @@
+                # Disable until the Twitter tokenizer is available.
+                rb.setEnabled(False)
+            contentArea.addWidget(rb)
+            self.group.addButton(rb, method)
+        self.group.buttonClicked.connect(self.group_button_clicked)
+
+        self.add_widget(self.contents)
+
+        # Restore the previous state, after starting off the layout.
+        self.restore_data(data)
+
+    def group_button_clicked(self):
+        self.tokenizer_method = self.group.checkedId()
+        self.notify_on_change()
+
+    def restore_data(self, data):
+        self.tokenizer_method = data.get('method')
+        b = self.group.button(self.tokenizer_method)
+        b.setChecked(True)
+
+    def export_data(self):
+        return {
+            'is_enabled': self.enabled,
+            'method': self.tokenizer_method,
+        }
+
+    def get_pp_setting(self):
+        return {
+            'use_twitter_tokenizer': self.tokenizer_values.get(
+                    self.tokenizer_method)
+        }
+
+
+class TransformationModule(PreprocessorModule):
+    DEFAULT_SETTINGS = {
+        'is_enabled': True,
+        'method': 0,
+    }
+
+    PorterStemmer, SnowballStemmer, Lemmatizer = 0, 1, 2
+    transformation_values = {
+        PorterStemmer: PS,
+        SnowballStemmer: SS,
+        Lemmatizer: LM,
+    }
+    transformation_names = {
+        PorterStemmer: 'Porter stemmer',
+        SnowballStemmer: 'Snowball stemmer',
+        Lemmatizer: 'Lemmatizer',
+    }
+
+    transformation_method = 0
+
+    def __init__(self, data):
+        if not data:
+            data = self.DEFAULT_SETTINGS
+        PreprocessorModule.__init__(self, 'Stemming', True,
+                                    data.get('is_enabled'))
+
+        self.contents = QWidget()
+        contentArea = QVBoxLayout()
+        contentArea.setContentsMargins(15, 10, 15, 10)
+        self.contents.setLayout(contentArea)
+
+        self.group = QButtonGroup(self, exclusive=True)
+        for method in [
+            self.PorterStemmer,
+            self.SnowballStemmer,
+            self.Lemmatizer
+        ]:
+            rb = QRadioButton(self, text=self.transformation_names[method])
+            contentArea.addWidget(rb)
+            self.group.addButton(rb, method)
+        self.group.buttonClicked.connect(self.group_button_clicked)
+
+        self.add_widget(self.contents)
+
+        # Restore the previous state, after starting off the layout.
+        self.restore_data(data)
+
+    def group_button_clicked(self):
+        self.transformation_method = self.group.checkedId()
+        self.notify_on_change()
+
+    def restore_data(self, data):
+        self.transformation_method = data.get('method')
+        b = self.group.button(self.transformation_method)
+        b.setChecked(True)
+
+    def export_data(self):
+        return {
+            'is_enabled': self.enabled,
+            'method': self.transformation_method,
+        }
+
+    def on_toggle(self):
```

Comment:
These `on_toggle` methods in `TransformationModule`, `CasingModule` and `FilteringModule` are an other reason why `self.content` should be moved into the parent class. Then all these modules would not need a copy of the same method, since showing/hiding of content area would be done in the parent class.